core-v1/src/main/scala/lens/VariantLens.scala

```diff
@@ -0,0 +1,269 @@
+      *
+      * TODO: Should there be any path information for conversion?
+      *       List => Map seems important information as values with duplicate keys could be dropped
+      */
+    def to[B](factory: Factory[E, B]): VariantLens[A, B] =
+      lens.copy(get = lens.get.andThen(factory.fromSpecific(_)))
+
+    /**
+      * Get the first element of this [[IterableOnce]], if it isn't empty.
+      */
+    def headOption: VariantLens[A, Option[E]] =
+      lens.copy(
+        path = lens.path.atHead,
+        get = lens.get.andThen(b => View.from[E](b).headOption),
+      )
+  }
+
+  implicit final class AsHListBuilder[A, L <: HList](private val lens: VariantLens[A, L]) extends AnyVal {
+
+    /**
+      * Convert the resulting [[HList]] into a tuple.
+      */
+    def tupled[T](implicit tupler: hlist.Tupler.Aux[L, T]): VariantLens[A, T] =
+      lens.copy(
+        path = lens.path,
+        get = lens.get.andThen(tupler.apply),
+      )
+  }
+}
+
+sealed trait VariantLensLowPriorityImplicits {
+
+  /**
+    * Wrap the result of this lens with an [[AsIterableBuilder]] for helper operations on [[IterableOnce]] types.
+    */
+  implicit def asIterable[A, B, C[x] <: IterableOnce[x], E](
+    lens: VariantLens[A, B],
+  )(implicit
+    ev: B <:< C[E],
+  ): VariantLens.AsIterableBuilder[A, C, E] =
+    new VariantLens.AsIterableBuilder(lens.as[C[E]])
+}
+
+/**
+  * A serializable lens for extracting a value from an object.
+  *
+  * @param path the serializable path to the value from the starting type to the value type
+  * @param get a function for extracting the field value
+  * @tparam A the starting type
+  * @tparam B the extracted value type
+  */
+final case class VariantLens[-A, +B](
+  path: DataPath,
+  get: A => B,
+) { outer =>
+
+  /**
+    * Composes another lens to run on the output of this lens.
+    *
+    * @note Because extension methods don't retain type information well,
+    *       this is still useful despite the [[Compose]] instance.
+    */
+  def andThen[C](lens: VariantLens[B, C]): VariantLens[A, C] = {
+    copy(
+      path = this.path ++ lens.path,
+      get = get.andThen(lens.get),
+    )
+  }
+
+  // Helpful for building a tuple with the lens itself. Identical to identity, but you don't have to specify the type.
+  def self: VariantLens[A, B] = this
+
+  /**
+    * Access a field from the object returned by this lens.
+    *
+    * Typically, you would not call this directly, but instead by calling the [[NamedLens.Selector.select]]
+    * extension method on this lens. The `.select` method uses macros to make sure that the `name` param is
+    * set properly.
+    *
+    * @param name the name of the field
+    * @param getter a function to extract the value of this field from the object
+    * TODO: Use HList to make this more safe, instead of relying on the macro to use this properly.
+    */
+  def field[C](
```

Comment:
I doubt I'll ever get around to this TODO. I should probably remove it. Just being careful and using the macro seems to be pretty safe. I doubt adding shapeless will do anything other than introduce difficulty without much benefit.